### ISSUE TITLE
How can I install 'z' in fish?

### ISSUE THREAD

**Issue by ManasN** [oh-my-fish/plugin-z#2]
Hello, since the rupa/z plugin comes in bash format, fish is not able to install it.
And hence pkg-z is not working, how can I install 'z' for fish?
Thanks.

**Comment by bpinto** [comment]
It depends on your OS, if you are using OS X 'brew install z".

It's likely that you didn't configure your script path as explained on the readme page, take a look and see how it goes.

**Comment by ManasN** [comment]
I'm using Linux Mint 17.2. I've double checked the path configuration, it's correct.

Case 1 : 
set -g Z_SCRIPT_PATH #path to rupa/z script, this line is added at the starting of config.fish file. Same error.

Case 2 : 
set -g Z_SCRIPT_PATH #path to pkg-z script, added at the beginning, same error.

**Comment by bpinto** [comment]
You need to replace #path with the real path, have you done this?

**Comment by ManasN** [comment]
Yes, ofcourse!
First I replaced it with the path of rupa/z script and got the same error.
Then I replaced it with pkg-z script and still getting the same error.
Thanks.

**Comment by bpinto** [comment]
Just to be clear, this is what I'm suggesting:

```
# Path to your oh-my-fish.
set -g OMF_PATH $HOME/.local/share/omf

# Path to your oh-my-fish configuration.
set -g OMF_CONFIG $HOME/.config/omf

### Configuration required to load oh-my-fish ###
# Note: Only add configurations that are required to be set before oh-my-fish is loaded.
# For common configurations, we advise you to add them to your $OMF_CONFIG/init.fish file or
# to create a custom plugin instead.
set -gx Z_SCRIPT_PATH /usr/local/etc/profile.d/z.sh

# Load oh-my-fish configuration.
source $OMF_PATH/init.fish
```

**Comment by ManasN** [comment]
It worked! I setting Z_SCRIPT_  at the wrong time. :+1: Thanks!

**Comment by jonscottclark** [comment]
Hey @bpinto, could you possibly clarify the installations instructions in the README to echo this advice? I had no idea where to find z.sh until I came here. Thanks!

**Comment by bpinto** [comment]
notify @derekstavis

**Comment by jonscottclark** [comment]
Thanks @bpinto, and does this repo need to be renamed to **plugin-z** ?

**Comment by bpinto** [comment]
It is a good idea, I can do this now as github will redirect automatically. Thanks @jonscottclark

**Comment by ffittschen** [comment]
@bpinto I seem to a similar problem.. I installed z using `brew install z` and modified my `config.fish`, but it doesn't work.

``` bash
# Path to Oh My Fish install.
set -gx OMF_PATH "/Users/ffittschen/.local/share/omf"

# Path to z
set -gx Z_SCRIPT_PATH /usr/local/etc/profile.d/z.sh

# Load oh-my-fish configuration.
source $OMF_PATH/init.fish
```

I checked, that `z` is executable, but it doesn't even create a `.z` file in my `$HOME` folder.

Do you have any suggestions, or should I create an issue at the rupa/z repo ?

**Comment by mathop** [comment]
@ffittschen Same here, just installed fish + omf + z in a fresh OSX and it doesn't work:

```
$ brew install z
$ omf install z
```

Added **Z_SCRIPT_PATH** into my **config.fish**:

> set -g Z_SCRIPT_PATH (brew --prefix)/etc/profile.d/z.sh

```
$ ls -lah $Z_SCRIPT_PATH
lrwxr-xr-x  1 matheus.pedroso  admin    37B Feb  3 13:45 /usr/local/etc/profile.d/z.sh -> ../../Cellar/z/1.9/etc/profile.d/z.sh
```

**Comment by bpinto** [comment]
I don't think it would create the `.z` folder for you automatically. I think you will have to run `mkdir ~/.z`. If you open a PR we could do this automatically on `install.fish`. So whenever this plugin is installed we create the directory.

---

The place where @ffittschen specified the path is correct as it happens before sourcing OMF.

```
# Path to z
set -gx Z_SCRIPT_PATH /usr/local/etc/profile.d/z.sh

# Load oh-my-fish configuration.
source $OMF_PATH/init.fish
```

Is yours the same @mathop?

**Comment by jonscottclark** [comment]
@ffittschen see https://github.com/rupa/z/issues/139, maybe?

**Comment by mathop** [comment]
@bpinto Yep, I set Z_SCRIPT_PATH before sourcing OMF.

About creating the directory:

https://github.com/rupa/z/blob/master/README

> Data is stored in $HOME/.z. This  can  be  overridden  by  setting  the
>        $_Z_DATA  environment variable. When initialized, z will raise an error
>        if this path is a directory, and not function correctly.

**Comment by mathop** [comment]
@bpinto I think this issue just happens after the commit 14f7d3e6ec6fa3791c938f010e8b861280529962.

It works fine till the commit 175582fe4364f0d7dcd370599380a76b6c2550de:

```
$ cd .local/share/omf/pkg/z
$ git checkout 175582fe4364f0d7dcd370599380a76b6c2550de
```

I could make it work with last commit just moving back the `z.check` function inside `init.fish`.

**Comment by leonmax** [comment]
@mathop you are right! it was broken after that commit!

z was there, just it writes nothing to $HOME/.z, and thus you can't jump without the history.

Something must be rotten in this evil commit [`14f7d3e`](https://github.com/oh-my-fish/plugin-z/commit/14f7d3e6ec6fa3791c938f010e8b861280529962)

**Comment by derekstavis** [comment]
Yes. @mathop is right. The issue is that `z.check` is an event handler:

``` fish
function z.check --on-variable PWD --description 'Setup z on directory change'
```

And fish does not evaluate `z.check.fish` until `z.check` is called. It's also possible to `source $path/functions/z.check.fish` in `init.fish` to keep the code separation.

**Comment by seyhaheng2** [comment]
why now i can't install oh my fish anymore please help

**Comment by alessandro-fazzi** [comment]
It's due to a DMCA occurring. The repo is anymore public at the moment

**Comment by derekstavis** [comment]
We will be back into business in few days. Sorry for the downtime and lack of communication.

**Comment by bobthecow** [comment]
The OMF repo has been restored. Thank you for your patience.

**Comment by tasmo** [comment]
Tried almost everything descripted here, nothing works.

**Comment by ma1f** [comment]
Any plans to fix this? Doesn't work and tried various suggestions described above to no avail.
Have `z` working just fine with bash.

> `# Path to z`
> `set -gx Z_SCRIPT_PATH /home/.../z.sh`

No effect, fish complains about `Please install 'z' first!`

> `# Load oh-my-fish configuration.`
> `source $OMF_PATH/init.fish`

causes infinite loop - likely due to plugins being loaded before config.fish

**Comment by cesards** [comment]
- 1 to this:

```
 ~/.l/s/o/p/z   master  fish                                                                 Tue  6 Sep 13:15:36 2016
Please install 'z' first!
Darwin MacBook-Pro.local 15.6.0 x86_64
13:15  up 3 days, 23:05, 2 users, load averages: 2.35 2.39 2.41
```

**Comment by scorphus** [comment]
Make sure `z` is installed:

```
»»»» brew info z
z: stable 1.9, HEAD
Tracks most-used directories to make cd smarter
https://github.com/rupa/z
/usr/local/Cellar/z/1.8 (4 files, 18.8K)
  Built from source on 2015-04-28 at 19:01:12
/usr/local/Cellar/z/1.9 (4 files, 19.6K) *
  Built from source on 2016-06-01 at 17:51:51
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/z.rb
==> Caveats
For Bash or Zsh, put something like this in your $HOME/.bashrc or $HOME/.zshrc:
  . `brew --prefix`/etc/profile.d/z.sh
```

We're in Fish, Bash's backquotes won't work here. We better use parenthesis:

```
»»»» echo (brew --prefix)/etc/profile.d/z.sh
/usr/local/etc/profile.d/z.sh
```

Voila! That's the script location! Make sure it's in `config.fish`:

```
»»»» grep z ~/.config/fish/config.fish
# plugin-z script path
set -g Z_SCRIPT_PATH /usr/local/etc/profile.d/z.sh
```

Make sure `plugin-z` is installed:

```
»»»» omf list | grep z
z
```

Make sure it's loaded:

```
»»»» functions z
function z --description 'jump around'
        cd (bash -c "source $Z_SCRIPT_PATH; _z $argv; echo \$PWD")
end
```

\o/

**Comment by scorphus** [comment]
Note: some `omf reload` here and there might be necessary.

**Comment by ma1f** [comment]
> make sure ... is installed

Did you actually read my comment? I clearly stated everything was installed, repeating this as an answer isn't a solution.

> fish complains about Please install 'z' first!

This is from the z-plugin, so yes the z-package is installed.
If you review the log below, you can see that the plugin is loaded before setting `Z_SCRIPT_PATH` in `config.fish`, and therefore `z` works if the fish shell is reloaded (as it is now set)

```
deadpool@VERTIGO:~$ fish
Please install 'z' first!
hello from config.fish
 ~ > z
 ~ > cd .ssh
 ~/.ssh > cd ~/
 ~ > z
 ~ > fish
hello from config.fish
 ~ > cd .ssh
 ~/.ssh > cd ~/
 ~ > z
common:    /home/deadpool/.ssh
4          /home/deadpool/.ssh
 ~ >
```

and config.fish

```
 ~ > cat ~/.config/fish/config.fish
#! ~/.config/fish/config.fish
set -gx fish_greeting ""
echo 'hello from config.fish'
# path to z
set -gx Z_SCRIPT_PATH /home/deadpool/z.sh
```

**Comment by scorphus** [comment]
> Did you actually read my comment?

Yes 😄 

> > fish complains about Please install 'z' first!

Let's have a look at [`plugin-z`'s `init` function](https://github.com/oh-my-fish/plugin-z/blob/02ed2f71ac362dfb96cf130e38deb9cad0db5c15/init.fish#L2):

``` fish
  if test -z $Z_SCRIPT_PATH
    echo "Please install 'z' first!"; return 1
  end
```

And at the section of [`test` manual](https://fishshell.com/docs/current/commands.html#test) regarding `-z`:

> ```
>    o -z STRING returns true if the length of STRING is zero.
> ```

According to the above, it looks like you haven't set `$Z_SCRIPT_PATH`. At least not early enough.

These might help:

``` fish
»»»» set -q Z_SCRIPT_PATH; and echo ok; or echo too bad
```

``` fish
»»»» test -n $Z_SCRIPT_PATH; and echo ok; or echo too bad
```

**Comment by ma1f** [comment]
> According to the above, it looks like you haven't set $Z_SCRIPT_PATH. At least not early enough.

Where would the correct place to set this if not in `~/.config/fish/config.fish` as you recommend?

```
 ~ > cat ~/.config/fish/config.fish
#! ~/.config/fish/config.fish
set -gx fish_greeting ""
echo 'hello from config.fish'
# path to z
set -gx Z_SCRIPT_PATH /home/deadpool/z.sh
```

You can see that the plugin is executed before `config.fish`

```
deadpool@VERTIGO:~$ fish
Please install 'z' first!
hello from config.fish
 ~ >
```

**Comment by bobthecow** [comment]
Try adding it to `~/.config/omf/before.init.fish` if that's not early enough.

**Comment by ma1f** [comment]
Cheers @bobthecow, solved it nicely.

Might be worth updating the README as this was from a clean install of both Ubuntu 14 and Ubuntu 16 and also appears to be affecting @cesards on OSX darwin.

**Comment by bobthecow** [comment]
I think it has to do with the update to Fish 2.3+ style `conf.d`. Before that, putting it in `~/.config/fish/config.fish` would have been enough.

**Comment by cesards** [comment]
Yep, I think updating the README would be an option. And just for everyone's sake:

as @bobthecow said, this is the solution:

```
$ touch .config/omf/before.init.fish # if you don't have that file created
$ vim .config/omf/before.init.fish
```

and add the following:

```
# plugin-z
set -g Z_SCRIPT_PATH /usr/local/etc/profile.d/z.sh
```

Cheers guys!

**Comment by mjobuda** [comment]
I'm sorry, but the docs of this soft suck. I don't have /usr/local/etc/profile.d/z.sh on my system, it is not described how to install, configure and use it. It is not even clear what this plugin does :)

**Comment by scorphus** [comment]
It is a wrapper for https://github.com/rupa/z

**Comment by sagebind** [comment]
This has now been fixed with #11. z is now bundled automatically with the plugin and does not require any annoying configuration to work.
